### PR TITLE
coverage: Unify `tests/coverage-map` and `tests/run-coverage` into `tests/coverage`

### PR DESCRIPTION
**Pull request by Zalathar**
Ever since the introduction of the `coverage-map` suite, it's been awkward to have to manage two separate coverage test directories containing dozens of mostly-identical files.

However, those two suites were separate for good reasons. They have very different requirements (since only one of them requires actually running the test program), running only one suite is noticeably faster than running both, and having separate suites allows them to be blessed separately if desired. So while unifying them was an obvious idea, actually doing so was non-trivial.

---

Nevertheless, this PR finds a way to merge the two suites into one directory while retaining almost all of the developer-experience benefits of having two suites. This required non-trivial implementations of `Step`, but the end result works very smoothly.

---

The first 5 commits are a copy of #117340, which has been closed in favour of this PR.

### CODE REVIEW

**Comment by Zalathar** [comment]
Some other approaches that I considered but ultimately rejected:

- The original idea was to just declare two different test suites in bootstrap that point to the same directory. However, bootstrap contains some assumptions that aren't compatible with this, so the easy approach doesn't actually work.
- An obvious idea is to fully unify the two suites/modes into one compiletest mode that runs both sets of checks. This carries several drawbacks:
  - A key advantage of the coverage-map tests is that they *don't* require the profiler runtime, which lets them run in more CI configurations (e.g. for PRs) and makes them easier to run locally. There's no way to reconcile that with the fact that run-coverage tests *do* require the profiler runtime, which is a major reason why the two suites were separate in the first place.
- A related idea I had was to hook into the compiletest code that splits a test into multiple revisions, and have it also split run-coverage tests into multiple *modes*.
  - This would have required adding several new features to compiletest to preserve important properties of the existing developer experience.
  - I did have a mostly-working implementation of this, but once I figured out how to do a similar thing at the bootstrap level with aliases, the bootstrap approach seemed like a clearly superior solution.

File: src/bootstrap/src/core/build_steps/test.rs [hunk 1]
```diff
@@ -1349,13 +1391,68 @@ host_test!(RunMakeFullDeps {
...
+    alias: "coverage-map",
+    default: true,
+    host: false,
+});
+coverage_test_alias!(RunCoverage {
+    mode: "run-coverage",
+    alias: "run-coverage",
+    default: true,
+    host: true,
 });
```
[the remark this answers is not in the document]
Comment (Zalathar) [reply src/bootstrap/src/core/build_steps/test.rs:1451]: Originally I though having them separate was clearer, but then I realized I can just call them `alias_and_mode` and get the best of both worlds.

**Comment by Zalathar** [comment]
I've also been thinking about renaming the `run-coverage` alias/mode to `coverage-run`, but I think that's too much scope creep for this PR. Better to keep the old name for now, and worry about changing it in a separate PR.

**Comment by Zalathar** [comment]
Found and fixed an additional papercut: Because compiletest was using the same output directory for both modes, they would interfere with each others' timestamp files.

I've fixed this by adding a special case to compiletest's output path for coverage test modes [(diff)](https://github.com/rust-lang/rust/compare/57d3f19b485e59c3f8be95e959dc2d9207a0e4bb..502eaaacc36615d647e51d688abba7f3ff2bd737).

**Comment by Zalathar** [comment]
Oh, I notice there's already a `Mode::disambiguator` method that serves a similar purpose; perhaps I should hook into that instead.

**Review by Swatinem — APPROVED** [review]

**Comment by Zalathar** [comment]
Looks like output normalization is missing paths on Windows.

Perhaps because we’re using `suite_path` in `--src-base`, which has a forward slash.

**Comment by Zalathar** [comment]
Updated to replace`/` in `suite_path` with `MAIN_SEPARATOR_STRING`, so that Windows should consistently see `\` as the path separator [(diff)](https://github.com/rust-lang/rust/compare/7d1035d204030f6e6ccced82410e20cf1c9614b8..d8df5d98c1d8d9d929aa73104ead1d49ce59f192).

I'll do a Windows test run on the side in #114917 to verify that this actually fixes the problem.

**Comment by Zalathar** [comment]
I'm a bit surprised that the path separator affects output-normalization of coverage tests, but not other tests (e.g. `tests/ui` runs fine).

Maybe there's some amount of input path normalization that is being performed by the compiler on its normal diagnostics, but coverage mappings embed the "raw" path that was passed on the command-line.

**Comment by Zalathar** [comment]
After some more investigation, I realised that the patch that switches to `suite_path` shouldn't be needed by this PR at all.

(It was needed by earlier prototypes, but in the final version the suite name and suite path always match; it's only the mode that  varies.)

While I still think it's abstractly “correct” to apply the fix, I'll drop it from this PR entirely.

**Comment by Zalathar** [comment]
I've resolved the Windows build issues by removing the patch that modifies `--src-base`, since it's not actually needed by the rest of this PR.

@rustbot ready

**Pull request merged**